File: ai_service/app/schemas/video_generation.py
"""
Schemas for AI Video Generation API.
"""
from __future__ import annotations

from typing import Optional, Dict, Any, List, Literal
from pydantic import BaseModel, Field


class ReferenceFileItem(BaseModel):
    """A reference file (image or PDF) attached to a generation request."""
    url: str = Field(..., description="Public S3 URL of the file")
    name: str = Field(..., description="Original filename (e.g., 'diagram.png')")
    type: Literal["image", "pdf"] = Field(..., description="File type: 'image' or 'pdf'")


# Content types supported by the generation pipeline
ContentTypeEnum = Literal[
    "VIDEO",              # Time-synced HTML overlays with audio (default)
    "QUIZ",               # Question-based assessments  
    "STORYBOOK",          # Page-by-page narratives
    "INTERACTIVE_GAME",   # Self-contained HTML games
    "PUZZLE_BOOK",        # Collection of puzzles
    "SIMULATION",         # Physics/economic sandboxes
    "FLASHCARDS",         # Spaced-repetition cards
    "MAP_EXPLORATION",    # Interactive SVG maps
    # New content types
    "WORKSHEET",          # Printable/interactive homework
    "CODE_PLAYGROUND",    # Interactive code exercises
    "TIMELINE",           # Chronological event visualization
    "CONVERSATION",       # Language learning dialogues
    "SLIDES"              # HTML-based presentation / PPT-style slide deck
]


class VideoGenerationRequest(BaseModel):
    """Request for generating AI video or interactive content."""
    
    prompt: str = Field(..., description="Text prompt for content generation")
    content_type: ContentTypeEnum = Field(
        default="VIDEO",
        description="Type of content to generate. Determines navigation mode and required libraries."
    )
    language: str = Field(default="English", description="Language for content (e.g., English, Spanish, French)")
    captions_enabled: bool = Field(default=True, description="Enable/disable captions (primarily for VIDEO type)")
    html_quality: str = Field(default="advanced", description="HTML quality mode: 'classic' (frames/animation only) or 'advanced' (all features)")
    video_id: Optional[str] = Field(default=None, description="Optional content ID (generated if not provided)")
    target_audience: str = Field(
        default="General/Adult", 
        description="Target audience for age-appropriate content. Examples: 'Class 3 (Ages 7-8)', 'Class 9-10 (Ages 14-15)', 'College/Adult'"
    )
    target_duration: str = Field(
        default="2-3 minutes", 
        description="Target duration for VIDEO type. For other types, controls content length."
    )
    institute_id: Optional[str] = Field(
        default=None,
        description="Institute identifier (optional, for logging/context)"
    )
    user_id: Optional[str] = Field(
        default=None,
        description="User identifier (optional, for logging/context)"
    )
    model: Optional[str] = Field(
        default=None,
        description="AI Model to use for generation (e.g. 'xiaomi/mimo-v2-flash:free')"
    )
    quality_tier: str = Field(
        default="ultra",
        description="Quality tier: 'free', 'standard', 'premium', 'ultra', 'super_ultra'. Controls two-pass script review, HTML validation, image prompt enhancement, crossfade transitions, kinetic text shots, and other quality features."
    )
    voice_gender: str = Field(
        default="female",
        description="Voice gender for TTS (VIDEO/STORYBOOK): 'male' or 'female'. Default is 'female'."
    )
    tts_provider: str = Field(
        default="standard",
        description="TTS tier: 'standard' (Microsoft Edge TTS, free) or 'premium' (Google Cloud / Sarvam AI). Premium auto-routes: Indian languages → Sarvam AI, global languages → Google Cloud TTS."
    )
    voice_id: Optional[str] = Field(
        default=None,
        description="Specific voice ID for premium TTS. For Sarvam (Indian): e.g. 'ritu', 'shubh', 'priya'. For Google: e.g. 'en-US-Journey-F'. If not provided, a default voice is chosen based on language and gender."
    )
    generate_avatar: bool = Field(
        default=False,
        description="Whether to generate a talking head avatar. Defaults to False."
    )
    avatar_image_url: Optional[str] = Field(
        default=None,
        description="URL of a face/portrait image for the speaking avatar. If not provided, a default teacher image is used. Only applies to VIDEO content type."
    )
    reference_files: Optional[List[ReferenceFileItem]] = Field(
        default=None,
        description="List of reference files (images/PDFs) to include in generation"
    )
    orientation: Literal["landscape", "portrait"] = Field(
        default="landscape",
        description="Video orientation: 'landscape' (1920x1080, 16:9) or 'portrait' (1080x1920, 9:16)"
    )
    visual_style: str = Field(
        default="standard",
        description=(
            "DEPRECATED — accepted for API back-compat but no longer gates behavior. "
            "The Director LLM now picks theme, background, and animation language "
            "per-shot based on content, and can shift styles across a long video's "
            "timeline. Kept on the request schema so existing clients don't break."
        )
    )
    sound_effects_enabled: bool = Field(
        default=True,
        description=(
            "Enable automatic sound effects (transitions, chimes, impacts) baked "
            "into the video timeline. When True (default at premium+), the Sound "
            "Planner derives cues from shot types, sync points, skill audio events "
            "and narration emphasis — no extra Director burden. When False, all "
            "cues are suppressed regardless of tier."
        )
    )

    class Config:
        json_schema_extra = {
            "example": {
                "prompt": "Explain the concept of quantum entanglement to a 5 year old",
                "content_type": "VIDEO",
                "language": "English",
                "captions_enabled": True,
                "html_quality": "advanced",
                "video_id": "quantum-entanglement-101",
                "target_audience": "Class 3 (Ages 7-8)",
                "target_duration": "5 minutes",
                "voice_gender": "female",
                "tts_provider": "standard",
                "voice_id": None,
                "avatar_image_url": None,
                "orientation": "landscape",
                "reference_files": [
                    {"url": "https://bucket.s3.amazonaws.com/file1.png", "name": "diagram.png", "type": "image"},
                    {"url": "https://bucket.s3.amazonaws.com/file2.pdf", "name": "notes.pdf", "type": "pdf"}
                ]
            }
        }


class VideoGenerationResumeRequest(BaseModel):
    """Request for resuming video generation from a checkpoint."""
    
    video_id: str = Field(..., description="Video ID to resume")
    generate_avatar: bool = Field(
        default=False,
        description="Whether to generate a talking head avatar. Defaults to False."
    )
    avatar_image_url: Optional[str] = Field(
        default=None,
        description="URL of a face/portrait image for the speaking avatar. If not provided, a default teacher image is used."
    )
    
    class Config:
        json_schema_extra = {
            "example": {
                "video_id": "quantum-entanglement-101",
                "avatar_image_url": None
            }
        }


class VideoStatusResponse(BaseModel):
    """Response for video/content generation status."""
    
    id: str
    video_id: str
    current_stage: str
    status: str
    content_type: str = Field(default="VIDEO", description="Content type: VIDEO, QUIZ, STORYBOOK, etc.")
    file_ids: Dict[str, Optional[str]]
    s3_urls: Dict[str, Optional[str]]
    prompt: Optional[str]
    language: str
    error_message: Optional[str]
    metadata: Dict[str, Any]
    created_at: Optional[str]
    updated_at: Optional[str]
    completed_at: Optional[str]
    
    class Config:
        json_schema_extra = {
            "example": {
                "id": "550e8400-e29b-41d4-a716-446655440000",
                "video_id": "quantum-entanglement-101",
                "current_stage": "HTML",
                "status": "COMPLETED",
                "content_type": "VIDEO",
                "file_ids": {
                    "script": "file-uuid-1",
                    "audio": "file-uuid-2",
                    "words": "file-uuid-3",
                    "timeline": "file-uuid-4"
                },
                "s3_urls": {
                    "script": "https://bucket.s3.amazonaws.com/ai-videos/quantum-entanglement-101/script/script.txt",
                    "audio": "https://bucket.s3.amazonaws.com/ai-videos/quantum-entanglement-101/audio/narration.mp3",
                    "words": "https://bucket.s3.amazonaws.com/ai-videos/quantum-entanglement-101/words/narration.words.json",
                    "timeline": "https://bucket.s3.amazonaws.com/ai-videos/quantum-entanglement-101/timeline/time_based_frame.json"
                },
                "prompt": "Explain the concept of quantum entanglement to a 5 year old",
                "language": "English",
                "error_message": None,
                "metadata": {},
                "created_at": "2024-01-15T10:30:00Z",
                "updated_at": "2024-01-15T10:35:00Z",
                "completed_at": "2024-01-15T10:35:00Z"
            }
        }


class VideoUrlsResponse(BaseModel):
    """Response for video HTML and audio URLs."""

    video_id: str
    html_url: Optional[str] = Field(None, description="URL to HTML timeline file (time_based_frame.json)")
    audio_url: Optional[str] = Field(None, description="URL to audio file (narration.mp3)")
    words_url: Optional[str] = Field(None, description="URL to time-synced words JSON for captions")
    avatar_url: Optional[str] = Field(None, description="URL to avatar talking-head video (MP4)")
    video_url: Optional[str] = Field(None, description="URL to rendered MP4 video (from render server)")
    status: str = Field(..., description="Current video generation status (PENDING, IN_PROGRESS, COMPLETED, FAILED, STALLED)")
    current_stage: str = Field(..., description="Current generation stage")
    updated_at: Optional[str] = Field(None, description="Last time the record was updated (ISO 8601)")
    error_message: Optional[str] = Field(None, description="Error message if generation failed or stalled")
    render_job_id: Optional[str] = Field(None, description="Active render job ID (for tracking render progress)")
    audio_tracks: Optional[List[Dict[str, Any]]] = Field(None, description="Extra audio tracks from meta.audio_tracks")

    class Config:
        json_schema_extra = {
            "example": {
                "video_id": "quantum-entanglement-101",
                "html_url": "https://bucket.s3.amazonaws.com/ai-videos/quantum-entanglement-101/timeline/time_based_frame.json",
                "audio_url": "https://bucket.s3.amazonaws.com/ai-videos/quantum-entanglement-101/audio/narration.mp3",
                "words_url": "https://bucket.s3.amazonaws.com/ai-videos/quantum-entanglement-101/audio/words.json",
                "avatar_url": None,
                "status": "COMPLETED",
                "current_stage": "HTML",
                "updated_at": "2024-01-15T10:35:00Z",
                "error_message": None
            }
        }


class RegenerateFrameRequest(BaseModel):
    """Request for regenerating a specific frame's HTML."""
    video_id: str = Field(..., description="Video ID")
    timestamp: float = Field(..., description="Timestamp of the frame in seconds")
    user_prompt: str = Field(..., description="User's instruction for modification")
    institute_id: Optional[str] = Field(None, description="Institute ID (optional)")


class RegenerateFrameResponse(BaseModel):
    """Response with new HTML content."""
    video_id: str
    frame_index: int
    timestamp: float
    original_html: str
    new_html: str


class UpdateFrameRequest(BaseModel):
    """Request for updating a specific frame's HTML."""
    video_id: str = Field(..., description="Video ID")
    frame_index: int = Field(..., description="Index of the frame to update")
    new_html: str = Field(..., description="New HTML content")


class AddFrameRequest(BaseModel):
    """Request for inserting a new frame into the timeline."""
    video_id: str = Field(..., description="Video ID")
    html: str = Field(..., description="HTML content for the new frame")
    # Time-driven fields (optional — omit for user_driven videos)
    in_time: Optional[float] = Field(None, description="Start time in seconds (time_driven)")
    exit_time: Optional[float] = Field(None, description="End time in seconds (time_driven)")
    # Optional explicit ID so the frontend can correlate the response
    entry_id: Optional[str] = Field(None, description="Client-generated entry ID (optional)")
    z: Optional[int] = Field(0, description="Z-index layer (0=base, 500+=overlay)")


class AddFrameResponse(BaseModel):
    status: str
    video_id: str
    entry_id: str
    frame_index: int
    message: str


# ── Audio track schemas ──────────────────────────────────────────────────────

class AudioTrackItem(BaseModel):
    """An extra audio track stored in meta.audio_tracks[]."""
    id: str = Field(..., description="Unique track ID (e.g. 'track-1')")
    label: str = Field(..., description="Display label (e.g. 'Background Music')")
    url: str = Field(..., description="Public S3 URL of the audio file")
    volume: float = Field(default=1.0, ge=0.0, le=2.0, description="Playback volume multiplier (0–2)")
    delay: float = Field(default=0.0, ge=0.0, description="Seconds to wait before starting")
    fade_in: float = Field(default=0.0, ge=0.0, description="Fade-in duration in seconds")
    fade_out: float = Field(default=0.0, ge=0.0, description="Fade-out duration in seconds")


class AddAudioTrackRequest(BaseModel):
    video_id: str = Field(..., description="Video ID")
    label: str = Field(..., description="Track display label")
    url: str = Field(..., description="Public S3 URL of the audio file")
    volume: float = Field(default=1.0, ge=0.0, le=2.0)
    delay: float = Field(default=0.0, ge=0.0)
    fade_in: float = Field(default=0.0, ge=0.0)
    fade_out: float = Field(default=0.0, ge=0.0)
    track_id: Optional[str] = Field(None, description="Client-provided track ID (auto-generated if absent)")


class UpdateAudioTrackRequest(BaseModel):
    video_id: str = Field(..., description="Video ID")
    track_id: str = Field(..., description="Track ID to update")
    label: Optional[str] = None
    url: Optional[str] = None
    volume: Optional[float] = Field(None, ge=0.0, le=2.0)
    delay: Optional[float] = Field(None, ge=0.0)
    fade_in: Optional[float] = Field(None, ge=0.0)
    fade_out: Optional[float] = Field(None, ge=0.0)


class DeleteAudioTrackRequest(BaseModel):
    video_id: str = Field(..., description="Video ID")
    track_id: str = Field(..., description="Track ID to delete")


class AudioTrackResponse(BaseModel):
    status: str
    video_id: str
    track_id: str
    message: str

File: src/routes/video-api-studio/-services/video-generation.ts
import { AI_SERVICE_BASE_URL } from '@/constants/urls';

export type VideoStage = 'PENDING' | 'SCRIPT' | 'TTS' | 'WORDS' | 'HTML' | 'RENDER';
export type VideoStatusType = 'PENDING' | 'IN_PROGRESS' | 'COMPLETED' | 'FAILED';

export type VoiceGender = 'female' | 'male';
export type TtsProvider = 'edge' | 'google';

/**
 * All supported content types from the API
 */
export type ContentType =
    | 'VIDEO' // Time-synced HTML overlays with audio (default)
    | 'QUIZ' // Question-based assessments
    | 'STORYBOOK' // Page-by-page narratives
    | 'INTERACTIVE_GAME' // Self-contained HTML games
    | 'PUZZLE_BOOK' // Collection of puzzles (crossword, word search)
    | 'SIMULATION' // Physics/economic sandboxes
    | 'FLASHCARDS' // Spaced-repetition cards
    | 'MAP_EXPLORATION' // Interactive SVG maps
    | 'WORKSHEET' // Printable/interactive homework
    | 'CODE_PLAYGROUND' // Interactive code exercises
    | 'TIMELINE' // Chronological event visualization
    | 'CONVERSATION'; // Language learning dialogues

/**
 * Navigation modes for content playback
 */
export type NavigationMode = 'time_driven' | 'user_driven' | 'self_contained';

/**
 * Mapping of content types to navigation modes
 */
export const CONTENT_TYPE_NAVIGATION: Record<ContentType, NavigationMode> = {
    VIDEO: 'time_driven',
    QUIZ: 'user_driven',
    STORYBOOK: 'user_driven',
    INTERACTIVE_GAME: 'self_contained',
    PUZZLE_BOOK: 'user_driven',
    SIMULATION: 'self_contained',
    FLASHCARDS: 'user_driven',
    MAP_EXPLORATION: 'user_driven',
    WORKSHEET: 'user_driven',
    CODE_PLAYGROUND: 'self_contained',
    TIMELINE: 'user_driven',
    CONVERSATION: 'user_driven',
};

/**
 * Content type options with labels and emojis for UI
 */
export const CONTENT_TYPES = [
    {
        value: 'VIDEO' as ContentType,
        label: '📹 Video',
        description: 'Time-synced HTML overlays with audio',
    },
    { value: 'QUIZ' as ContentType, label: '❓ Quiz', description: 'Question-based assessments' },
    {
        value: 'STORYBOOK' as ContentType,
        label: '📚 Storybook',
        description: 'Page-by-page narratives',
    },
    {
        value: 'INTERACTIVE_GAME' as ContentType,
        label: '🎮 Interactive Game',
        description: 'Self-contained HTML games',
    },
    {
        value: 'PUZZLE_BOOK' as ContentType,
        label: '🧩 Puzzle Book',
        description: 'Crossword, word search puzzles',
    },
    {
        value: 'SIMULATION' as ContentType,
        label: '🔬 Simulation',
        description: 'Physics/science sandboxes',
    },
    {
        value: 'FLASHCARDS' as ContentType,
        label: '📇 Flashcards',
        description: 'Spaced-repetition cards',
    },
    {
        value: 'MAP_EXPLORATION' as ContentType,
        label: '🗺️ Map Exploration',
        description: 'Interactive SVG maps',
    },
    {
        value: 'WORKSHEET' as ContentType,
        label: '📝 Worksheet',
        description: 'Printable/interactive homework',
    },
    {
        value: 'CODE_PLAYGROUND' as ContentType,
        label: '💻 Code Playground',
        description: 'Interactive code exercises',
    },
    {
        value: 'TIMELINE' as ContentType,
        label: '⏳ Timeline',
        description: 'Chronological event visualization',
    },
    {
        value: 'CONVERSATION' as ContentType,
        label: '💬 Conversation',
        description: 'Language learning dialogues',
    },
] as const;

export interface GenerateVideoRequest {
    prompt: string;
    content_type?: ContentType; // NEW: Default "VIDEO"
    language: string;
    voice_gender: VoiceGender;
    tts_provider: TtsProvider;
    captions_enabled: boolean;
    html_quality: 'classic' | 'advanced';
    target_audience: string;
    target_duration: string;
    model: string;
    video_id?: string; // Optional: auto-generated if not provided
}

export interface ProgressEvent {
    type: 'progress';
    stage: VideoStage;
    message: string;
    percentage: number;
    video_id: string;
    content_type?: ContentType; // NEW: Included in events
    files?: {
        script?: { file_id: string; s3_url: string };
        audio?: { file_id: string; s3_url: string };
        words?: { file_id: string; s3_url: string };
        timeline?: { file_id: string; s3_url: string };
        video?: { file_id: string; s3_url: string };
    };
}

export interface CompletedEvent {
    type: 'completed';
    video_id?: string;
    content_type?: ContentType; // NEW: Included in events
    files: {
        video?: string;
        script?: string;
        audio?: string;
        timeline?: string;
        words?: string;
    };
    percentage: number;
}

export interface InfoEvent {
    type: 'info';
    message: string;
    video_id?: string;
    content_type?: ContentType;
}

export interface ErrorEvent {
    type: 'error';
    message: string;
    stage?: VideoStage;
    video_id?: string;
}

export type SSEEvent = ProgressEvent | CompletedEvent | InfoEvent | ErrorEvent;

export interface VideoUrls {
    video_id: string;
    html_url: string;
    audio_url: string;
    words_url: string;
    current_stage: VideoStage;
}

export interface VideoStatusResponse {
    id: string;
    video_id: string;
    current_stage: VideoStage;
    status: VideoStatusType;
    s3_urls: {
        script?: string;
        audio?: string;
        video?: string;
    };
    created_at: string;
}

export interface HistoryItem {
    id: string;
    video_id: string;
    prompt: string;
    content_type: ContentType; // NEW: Track content type
    status: 'pending' | 'generating' | 'completed' | 'failed';
    stage: VideoStage;
    created_at: string;
    html_url?: string;
    audio_url?: string;
    video_url?: string;
    timeline_url?: string;
    words_url?: string;
    options: Omit<GenerateVideoRequest, 'prompt'>;
}

const HISTORY_STORAGE_KEY = 'vacademy_video_generation_history';

export const AVAILABLE_MODELS = [
    // Free Models
    { id: 'xiaomi/mimo-v2-flash:free', name: 'MIMO v2 Flash (Free)', tier: 'free' },
    { id: 'mistralai/devstral-2512:free', name: 'Devstral 2512 (Free)', tier: 'free' },
    { id: 'nvidia/nemotron-3-nano-30b-a3b:free', name: 'Nemotron 3 Nano (Free)', tier: 'free' },
    { id: 'google/gemini-2.0-flash-exp:free', name: 'Gemini 2.0 Flash Exp (Free)', tier: 'free' },
    // OpenAI Models
    { id: 'openai/gpt-4o-mini', name: 'GPT-4o Mini', tier: 'openai' },
    { id: 'openai/gpt-4o', name: 'GPT-4o', tier: 'openai' },
    { id: 'openai/gpt-4-turbo', name: 'GPT-4 Turbo', tier: 'openai' },
    { id: 'openai/gpt-3.5-turbo', name: 'GPT-3.5 Turbo', tier: 'openai' },
    // Gemini Models
    { id: 'google/gemini-2.5-pro', name: 'Gemini 2.5 Pro', tier: 'gemini' },
    { id: 'google/gemini-2.5-flash', name: 'Gemini 2.5 Flash', tier: 'gemini' },
    { id: 'google/gemini-2.5-flash-image', name: 'Gemini 2.5 Flash Image', tier: 'gemini' },
    { id: 'google/gemini-3-flash-preview', name: 'Gemini 3 Flash Preview', tier: 'gemini' },
    { id: 'google/gemini-2.5-flash-lite', name: 'Gemini 2.5 Flash Lite', tier: 'gemini' },
    { id: 'google/gemini-3-pro-preview', name: 'Gemini 3 Pro Preview', tier: 'gemini' },
    // OpenRouter / Other Paid Models
    { id: 'anthropic/claude-3.5-sonnet', name: 'Claude 3.5 Sonnet', tier: 'paid' },
    { id: 'meta-llama/llama-3.1-70b-instruct', name: 'Llama 3.1 70B', tier: 'paid' },
    { id: 'anthropic/claude-sonnet-4.5', name: 'Claude Sonnet 4.5', tier: 'paid' },
    { id: 'x-ai/grok-code-fast-1', name: 'Grok Code Fast 1', tier: 'paid' },
    { id: 'deepseek/deepseek-v3.2', name: 'DeepSeek V3.2', tier: 'paid' },
    { id: 'anthropic/claude-opus-4.5', name: 'Claude Opus 4.5', tier: 'paid' },
    { id: 'x-ai/grok-4.1-fast', name: 'Grok 4.1 Fast', tier: 'paid' },
] as const;

export const LANGUAGES = [
    // Global
    { value: 'English (US)', label: 'English (US)', group: 'Global' },
    { value: 'English (UK)', label: 'English (UK)', group: 'Global' },
    { value: 'Spanish', label: 'Spanish', group: 'Global' },
    { value: 'French', label: 'French', group: 'Global' },
    { value: 'German', label: 'German', group: 'Global' },
    { value: 'Japanese', label: 'Japanese', group: 'Global' },
    { value: 'Chinese', label: 'Chinese', group: 'Global' },
    // Indian
    { value: 'English (India)', label: 'English (India)', group: 'Indian' },
    { value: 'Hindi', label: 'Hindi', group: 'Indian' },
    { value: 'Bengali', label: 'Bengali', group: 'Indian' },
    { value: 'Tamil', label: 'Tamil', group: 'Indian' },
    { value: 'Telugu', label: 'Telugu', group: 'Indian' },
    { value: 'Marathi', label: 'Marathi', group: 'Indian' },
    { value: 'Kannada', label: 'Kannada', group: 'Indian' },
    { value: 'Gujarati', label: 'Gujarati', group: 'Indian' },
    { value: 'Malayalam', label: 'Malayalam', group: 'Indian' },
] as const;

export const VOICE_GENDERS = [
    { value: 'female', label: 'Female' },
    { value: 'male', label: 'Male' },
] as const;

export const TTS_PROVIDERS = [
    { value: 'edge', label: 'Standard (Free)' },
    { value: 'google', label: 'Premium (Google Cloud)' },
] as const;

export const TARGET_AUDIENCES = [
    'Class 1-2 (Ages 6-7)',
    'Class 3-4 (Ages 8-9)',
    'Class 5 (Ages 10-11)',
    'Class 6-8 (Ages 11-14)',
    'Class 9-10 (Ages 14-16)',
    'Class 11-12 (Ages 16-18)',
    'Undergraduate',
    'Graduate/Professional',
    'General/Adult',
];

export const TARGET_DURATIONS = [
    '30 seconds',
    '1 minute',
    '2-3 minutes',
    '5 minutes',
    '10 minutes',
];

export const DEFAULT_OPTIONS: Omit<GenerateVideoRequest, 'prompt'> = {
    content_type: 'VIDEO',
    language: 'English (US)',
    voice_gender: 'female',
    tts_provider: 'edge',
    captions_enabled: true,
    html_quality: 'advanced',
    target_audience: 'Class 5 (Ages 10-11)',
    target_duration: '2-3 minutes',
    model: 'xiaomi/mimo-v2-flash:free',
};

export function generateVideoId(): string {
    return `vid_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
}

/**
 * Response metadata from generation API
 */
export interface GenerationResponse {
    videoId: string;
    contentType: ContentType;
    abort: () => void;
}

/**
 * Generate content (video, quiz, storybook, etc.)
 */
export function generateVideo(
    request: GenerateVideoRequest,
    apiKey: string,
    onProgress: (event: SSEEvent) => void,
    onError: (error: Error) => void,
    onHeadersReceived?: (headers: { videoId: string; contentType: ContentType }) => void
): GenerationResponse {
    const videoId = request.video_id || generateVideoId();
    const controller = new AbortController();
    const contentType = request.content_type || 'VIDEO';

    const body = {
        ...request,
        video_id: videoId,
        content_type: contentType,
    };

    fetch(`${AI_SERVICE_BASE_URL}/external/video/v1/generate`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'X-Institute-Key': apiKey,
        },
        body: JSON.stringify(body),
        signal: controller.signal,
    })
        .then(async (response) => {
            if (!response.ok) {
                const errorText = await response.text().catch(() => response.statusText);
                throw new Error(`HTTP ${response.status}: ${errorText}`);
            }

            // Read content type from response headers
            const responseVideoId = response.headers.get('X-Video-ID') || videoId;
            const responseContentType =
                (response.headers.get('X-Content-Type') as ContentType) || contentType;

            // Notify caller of headers
            if (onHeadersReceived) {
                onHeadersReceived({
                    videoId: responseVideoId,
                    contentType: responseContentType,
                });
            }

            const reader = response.body?.getReader();
            if (!reader) {
                throw new Error('No response body');
            }

            const decoder = new TextDecoder();
            let buffer = '';

            // eslint-disable-next-line no-constant-condition
            while (true) {
                const { done, value } = await reader.read();
                if (done) break;

                buffer += decoder.decode(value, { stream: true });
                const lines = buffer.split('\n');
                buffer = lines.pop() || '';

                for (const line of lines) {
                    if (line.startsWith('data: ')) {
                        try {
                            let jsonStr = line.slice(6).trim();
                            // Handle Python-style single quotes by converting to double quotes
                            jsonStr = jsonStr
                                .replace(/'/g, '"')
                                .replace(/None/g, 'null')
                                .replace(/True/g, 'true')
                                .replace(/False/g, 'false');
                            const data = JSON.parse(jsonStr) as SSEEvent;
                            onProgress(data);
                        } catch (e) {
                            console.warn('SSE parse error:', e, 'Line:', line);
                        }
                    }
                }
            }
        })
        .catch((error) => {
            if (error.name !== 'AbortError') {
                onError(error);
            }
        });

    return {
        abort: () => controller.abort(),
        videoId,
        contentType,
    };
}

export async function getVideoUrls(videoId: string, apiKey: string): Promise<VideoUrls> {
    const response = await fetch(`${AI_SERVICE_BASE_URL}/external/video/v1/urls/${videoId}`, {
        headers: {
            'X-Institute-Key': apiKey,
        },
    });

    if (!response.ok) {
        throw new Error(`Failed to get video URLs: ${response.statusText}`);
    }

    return response.json();
}

export async function getVideoStatus(
    videoId: string,
    apiKey: string
): Promise<VideoStatusResponse> {
    const response = await fetch(`${AI_SERVICE_BASE_URL}/external/video/v1/status/${videoId}`, {
        headers: {
            'X-Institute-Key': apiKey,
        },
    });

    if (!response.ok) {
        throw new Error(`Failed to get video status: ${response.statusText}`);
    }

    return response.json();
}

export function getHistory(): HistoryItem[] {
    try {
        const stored = localStorage.getItem(HISTORY_STORAGE_KEY);
        return stored ? JSON.parse(stored) : [];
    } catch {
        return [];
    }
}

export function saveToHistory(item: HistoryItem): void {
    const history = getHistory();
    const existingIndex = history.findIndex((h) => h.video_id === item.video_id);

    if (existingIndex >= 0) {
        history[existingIndex] = item;
    } else {
        history.unshift(item);
    }

    // Keep only last 50 items
    const trimmed = history.slice(0, 50);
    localStorage.setItem(HISTORY_STORAGE_KEY, JSON.stringify(trimmed));
}

export function deleteFromHistory(videoId: string): void {
    const history = getHistory().filter((h) => h.video_id !== videoId);
    localStorage.setItem(HISTORY_STORAGE_KEY, JSON.stringify(history));
}

export function clearHistory(): void {
    localStorage.removeItem(HISTORY_STORAGE_KEY);
}

/**
 * Get navigation mode for a content type
 */
export function getNavigationMode(contentType: ContentType): NavigationMode {
    return CONTENT_TYPE_NAVIGATION[contentType] || 'time_driven';
}

/**
 * Check if content type requires audio
 */
export function requiresAudio(contentType: ContentType): boolean {
    return getNavigationMode(contentType) === 'time_driven';
}

/**
 * Get content type label for display
 */
export function getContentTypeLabel(contentType: ContentType): string {
    const found = CONTENT_TYPES.find((ct) => ct.value === contentType);
    return found?.label || contentType;
}
